sms_toolkit/utils/encoding.py

```diff
@@ -26,10 +31,20 @@ def convert_to_unicode_characters(string):
 
 
 def encode_unicode_character_to_utf16(character):
-    if len(character) > 1:
-        return flatten(map(encode_unicode_character_to_utf16, character))
-
-    return [((0xff00 & ord(character)) >> 8), 0x00ff & ord(character)]
+    """
+    Encode the given unicode code points into utf-16 and return a list a of bytes. Note that for code-points within
+    the BMP the number of bytes would be 2, but for code-points not in BMP, the number of bytes will be 4.
+
+    Args:
+        character: A unicode code-point
+
+    Returns:
+        A list of the bytes
+    """
+    if is_python_3():
+        return [c for c in character.encode('utf-16-be')]
+    else:
+        return [ord(c) for c in character.encode('utf-16-be')]
```

Comment:
<details>
<summary>Why is this change required? </summary>
<br />
<p>
The previous approach had one flaw - using `len`. Depending on whether the python interpreter build is narrow or wide (or in < py3.3 where the string encoding is based off of the highest code-point), the behaviour of len changes if the input is a code-point that's greater than what can be fit in 2 bytes (UTF-16 encodes them in 2 2-byte pairs, which we need to detect, since UCS-2 has a fixed 2 bytes limit).

In the narrow build - py2.7.12 (u'\U0001f600' is the unicode code point for 😀) - 
``` python 
>>> len(u'\U0001f600')
2
```

In wide build - py 2.7.12 
``` python 
>>> len(u'\U0001f600')
1
```

In python 3.5 (No need for u string prefix)- 
``` python 
>>> len('\U0001f600')
1
```
The issue here is that we want the len to be 2 everywhere. So we need to encode it to utf-16 explicitly.
</p>
</details>

<details>
<summary>Why the <code>be</code> in <code>utf-16-be</code> though? </summary>
<br />
<p>
Encodings can be done using big endian or little endian form (for the given 2 bytes one will have the higher byte first and the other will have the lower byte first). If we don't explicitly state that, then will be 2 additional bytes in the output which will be used to explicitly say what the endianess the method has used. That information is not required and We can directly tell the interpreter to use big-endian when encoding.
</p>
</details>

<details>
<summary>Why do we need to handle py2 and py3 separately?</summary>
<br />
<p>
str.encode behaves differently in python 2 and python 3.
In python 2 encode returns an 8-bit string version of the Unicode string, encoded in the requested encoding (<a href="https://docs.python.org/2/howto/unicode.html">Source</a>).
In python 3, encode returns an encoded version of the string as a bytes object. (<a href="https://docs.python.org/3/library/stdtypes.html#str.encode">Source</a>)

We basically want the list of byte values for the given character here. In python 3, we directly get it by iterating through every entry of encode. In py2, since the value is a string version, we can get the value by <code>ord</code>ing it.
</p>
</details>

<details>
<summary>But the <a href="https://github.com/chadselph/smssplit/blob/gh-pages/js/smssplit.js#L48-L56">original implementation</a> that the old code is based off of, seems to work well. How is that?</summary>
<br />
<p>
Yep, that'll always work because JS stores strings in 16-bits. It seems to be using UTF-16 and UCS-2 at different places, but a string character is 16 bits. <a href="https://mathiasbynens.be/notes/javascript-encoding">Source</a>. <code>"😀".length</code> always seems to return 2. So, since surrogate pairs are explicitly returned as 2 separate entries, that code will always work. Since we have to support different builds, we need to be more careful.
</p>
</details>

Reply:
This is super thorough, great work here.